HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift

```diff
@@ -15,45 +15,269 @@
+    let heartbeatPayload = controller.flush()
+
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01", "2021-11-02"]
+          }
+        ]
+      }
+      """
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+  }
+
+  func testDoNotLogDuplicate() throws {
+    // Given
+    let clock = SystemClock(date: date)
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { clock.date }
+    )
+
+    // When
+    controller.log("dummy_agent")
+    controller.log("dummy_agent")
+
+    // Then
+    let heartbeatPayload = controller.flush()
+
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+  }
+
+  func testDoNotLogDuplicateAfterFlushing() throws {
+    // Given
+    let clock = SystemClock(date: date)
+
+    let controller = HeartbeatController(
+      storage: HeartbeatStorageFake(),
+      dateProvider: { clock.date }
+    )
+
     // When
-    controller.log(#function)
+    controller.log("dummy_agent")
+    let heartbeatPayload = controller.flush()
+    controller.log("dummy_agent")
+
     // Then
-    XCTAssertNotNil(controller.flush())
-    XCTAssertNil(controller.flush())
+    assertEqualPayloadStrings(
+      heartbeatPayload.headerValue(),
+      """
+      {
+        "version": 0,
+        "payload": [
+          {
+            "agent": "dummy_agent",
+            "dates": ["2021-11-01"]
+          }
+        ]
+      }
+      """
+    )
+
+    // Below assertion asserts that duplicate was not logged again.
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
   }
 }
 
+func assertHeartbeatControllerFlushesEmptyPayload(_ controller: HeartbeatController) {
+  XCTAssertEqual(controller.flush().headerValue(), "")
+}
+
 // MARK: - Fakes
 
-private extension HeartbeatControllerTests {
+extension HeartbeatControllerTests {
   class HeartbeatStorageFake: HeartbeatStorageProtocol {
     private var heartbeatInfo: HeartbeatInfo?
 
-    func offer(_ heartbeat: Heartbeat) {
-      heartbeatInfo = HeartbeatInfo(capacity: 1)
-      heartbeatInfo!.offer(heartbeat)
+    func async(_ transform: @escaping HeartbeatInfoTransform) {
+      heartbeatInfo = transform(heartbeatInfo)
     }
 
-    func flush() -> HeartbeatInfo? {
-      let flushed = heartbeatInfo
-      heartbeatInfo = nil
-      return flushed
+    func getAndReset(using transform: HeartbeatInfoTransform?) throws -> HeartbeatInfo? {
+      let oldHeartbeatInfo = heartbeatInfo
+      heartbeatInfo = transform?(heartbeatInfo)
+      return oldHeartbeatInfo
     }
   }
+
+  /// Simulates the device system time.
+  class SystemClock {
```

Comment:
This name associates with a real system clock implementation for me. I would suggest updating the name with more reference to its intended use, e.g. `TestClock`. Also it seems it could be just a private extension on `Date` type. WDYT?

Reply:
Reworked in e9d0361. This led to some nice cleanup, see comment below.